pkg/controller/syncmachineset/syncmachineset_controller.go

```diff
@@ -0,0 +1,563 @@
+	clusterInformer informers.ClusterInformer,
+	kubeClient kubeclientset.Interface,
+	clusteroperatorClient coclient.Interface,
+) *Controller {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	// TODO: remove the wrapper when every clients have moved to use the clientset.
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(kubeClient.CoreV1().RESTClient()).Events("")})
+
+	if kubeClient != nil && kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("clusteroperator_syncmachineset_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	logger := log.WithField("controller", controllerLogName)
+
+	c := &Controller{
+		client:     clusteroperatorClient,
+		kubeClient: kubeClient,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "syncmachineset"),
+		logger:     logger,
+	}
+
+	machineSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addMachineSet,
+		UpdateFunc: c.updateMachineSet,
+		// TODO: delete
+	})
+	c.machineSetsLister = machineSetInformer.Lister()
+	c.machineSetsSynced = machineSetInformer.Informer().HasSynced
+
+	c.clusterLister = clusterInformer.Lister()
+
+	c.syncHandler = c.syncMachineSet
+	c.enqueueMachineSet = c.enqueue
+	c.buildClients = c.buildRemoteClusterClients
+
+	return c
+}
+
+// Controller manages syncing cluster and machineset objects with a
+// remote cluster API
+type Controller struct {
+	client     coclient.Interface
+	kubeClient kubeclientset.Interface
+
+	// To allow injection of syncMachineSet for testing.
+	syncHandler func(queuedMS queuedMachineSet) error
+
+	// Used for unit testing
+	enqueueMachineSet func(machineSet *cov1.MachineSet)
+
+	buildClients func(cluster *cov1.MachineSet) (clusterapiclient.Interface, kubeclientset.Interface, error)
+
+	// machineSetsLister is able to list/get machine sets and is
+	// populated by the shared informer passed to NewController.
+	machineSetsLister lister.MachineSetLister
+	// machineSetsSynced returns true if the machine set shared
+	// informer has been synced at least once.  Added as a member
+	// to the struct to allow injection for testing.
+	machineSetsSynced cache.InformerSynced
+
+	// clusterLister is able to list/get clusters and is populated
+	// by the shared informer passed to NewController.
+	clusterLister lister.ClusterLister
+	// clusterSynced returns true if the cluster shared informer
+	// has been synced at least once. Added as a member to the
+	// struct to allow injection for testing.
+	clusterSynced cache.InformerSynced
+
+	// Machines that need to be synced
+	queue workqueue.RateLimitingInterface
+
+	logger log.FieldLogger
+}
+
+// queuedMachineSet represents data we need to process a machine set in the queue. This includes
+// data we might need to process deleted objects we can no longer lookup.
+type queuedMachineSet struct {
```

Comment:
What is the purpose of this? I don't see `clusterName` used anywhere.

Reply:
Let's remove the `queuedMachineSet` struct entirely. Simply queue the keys. That will keep it consistent with how other controllers work.